gemfire-core/src/main/java/com/gemstone/gemfire/HeapDumper.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2010-2015 Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package com.gemstone.gemfire;
+
+import com.sun.management.HotSpotDiagnosticMXBean;
+
+import javax.management.MBeanServer;
+import java.lang.management.ManagementFactory;
+
+/**
+ * This class has been picked up from the following Oracle blog
+ * https://blogs.oracle.com/sundararajan/programmatically-dumping-heap-from-java-applications
+ */
+public class HeapDumper {
+    // This is the name of the HotSpot Diagnostic MBean
+    private static final String HOTSPOT_BEAN_NAME =
+            "com.sun.management:type=HotSpotDiagnostic";
+    // field to store the hotspot diagnostic MBean
+    private static volatile HotSpotDiagnosticMXBean hotspotMBean;
+
+    static void dumpHeap(boolean live) {
+        boolean dump = Boolean.getBoolean("gemfire.dumpheap");
+        if (!dump) {
+            return;
+        }
+        String fileName = "heap" + System.currentTimeMillis() + ".bin";
```

Comment:
Can you add protecting this so that only one heap dump happens in some time duration? Perhaps ForceDisconnect may not be invoked from multiple threads, but as a general API it is better to have the contract that this will not dump multiple times close together. For stack dumps there was a time limit e.g. make this method synchronized, record this System.currentTimeMillis() in a static long, then check against it that previous one should not be too close like < 3 seconds else skip the dump.